metadata/src/main/java/org/apache/kafka/image/publisher/BrokerRegistrationTracker.java

```diff
@@ -0,0 +1,135 @@
+    ) {
+        this.log = new LogContext("[BrokerRegistrationTracker id=" + id + "] ").
+            logger(BrokerRegistrationTracker.class);
+        this.id = id;
+        this.refreshRegistrationCallback = refreshRegistrationCallback;
+    }
+
+    @Override
+    public String name() {
+        return "BrokerRegistrationTracker(id=" + id + ")";
+    }
+
+    @Override
+    public void onMetadataUpdate(
+        MetadataDelta delta,
+        MetadataImage newImage,
+        LoaderManifest manifest
+    ) {
+        boolean checkBrokerRegistration = false;
+        if (delta.featuresDelta() != null) {
+            if (delta.metadataVersionChanged().isPresent()) {
+                if (log.isTraceEnabled()) {
+                    log.trace("Metadata version change is present: {}",
+                        delta.metadataVersionChanged());
+                }
+                checkBrokerRegistration = true;
+            }
+        }
+        if (delta.clusterDelta() != null) {
+            if (delta.clusterDelta().changedBrokers().get(id) != null) {
+                if (log.isTraceEnabled()) {
+                    log.trace("Broker change is present: {}",
+                        delta.clusterDelta().changedBrokers().get(id));
+                }
+                checkBrokerRegistration = true;
+            }
+        }
+        if (checkBrokerRegistration) {
+            if (brokerRegistrationNeedsRefresh(newImage.features().metadataVersion(),
+                    delta.clusterDelta().broker(id))) {
+                refreshRegistrationCallback.run();
+            }
+        }
+    }
+
+    /**
+     * Check if the current broker registration needs to be refreshed.
+     *
+     * @param registration  The current broker registration, or null if there is none.
+     * @return              True only if we should refresh.
+     */
+    boolean brokerRegistrationNeedsRefresh(
+        MetadataVersion metadataVersion,
+        BrokerRegistration registration
+    ) {
+        // If there is no existing registration, the BrokerLifecycleManager must still be sending it.
+        // So we don't need to do anything yet.
+        if (registration == null) {
+            log.debug("No current broker registration to check.");
+            return false;
+        }
+        // Check to see if the directory list has changed.  Note that this check could certainly be
+        // triggered spuriously. For example, if the broker's directory list has been changed in the
+        // past, and we are in the process of replaying that change log, we will end up here.
+        // That's fine because resending the broker registration does not cause any problems. And,
+        // of course, as soon as a snapshot is made, we will no longer need to worry about those
+        // old metadata log entries being replayed on startup.
+        if (metadataVersion.isAtLeast(MetadataVersion.IBP_3_7_IV2) &&
+                registration.directories().isEmpty()) {
```

Comment:
Should we also check the broker configuration? If the broker isn't configured with multiple log directories there's no need to re-register. Perhaps we shouldn't even setup this metadata listener if there's only one configured log dir.

Reply:
> If metadataVersion.isAtLeast(MetadataVersion.IBP_3_7_IV2) evaluates to false, then isn't registration.directories().isEmpty() always true?

No. We can get to this point in the code just because of a metadata version change. For example, from 3.7-IV1 to 3.7-IV2. In that case `registration.directories()` would not be empty and we would not want to do anything.

Reply:
> Should we also check the broker configuration? If the broker isn't configured with multiple log directories there's no need to re-register. Perhaps we shouldn't even setup this metadata listener if there's only one configured log dir.

broker IDs do have benefits for non-JBOD brokers. One benefit is knowing when a directory has been re-initialized. In any case, the impact of resending the registration in this specific case is pretty small.

Reply:
I don't think we add directories to registration records pre 3.7-IV2, so if we're bumping the MV from 3.7-IV1, how does `registration.directories()` get populated before a broker re-registration?

https://github.com/apache/kafka/blob/8d11d9579503426edfaeae791ec4bb212da37ad2/metadata/src/main/java/org/apache/kafka/controller/ClusterControlManager.java#L422-L424

Reply:
Sorry, I didn't express myself well. I agree that `registration.directories()` will indeed be empty if we're coming from `3.7-IV1` or earlier. But, we still need to check the current MV in this "if" statement since we do not want to resend the registration unless the MV is `3.7-IV2` or higher AND the registration does not contain directories.

Reply:
I see, the point is to trigger the re-registration only once, and empty directories signals that hasn't happened yet.